README.md
# LunaChess
luna: a UCI-compatible chess engine in c++, written by ori yonay :)

Special thanks / this project is highly inspired by:
- Jay Honnold (Berserk author)
- Finn Eggers (Koivisto author)
- Andrew Grant (Ethereal author)
- Manik Charan (Teki author)
- Oleg Smirnov (Zevra author)
- Jon Dart (Arasan/Fathom author)


extras/changelog.txt
starting rigorous testing of all features:
1) +30cp bonus for bishop pair: failed
2) +20cp bonus for bishop pair:
      Score of luna_dev vs luna_main: 2356 - 2195 - 1296  [0.514] 5847
    ...      luna_dev playing White: 1187 - 1093 - 644  [0.516] 2924
    ...      luna_dev playing Black: 1169 - 1102 - 652  [0.511] 2923
    ...      White vs Black: 2289 - 2262 - 1296  [0.502] 5847
    Elo difference: 9.6 +/- 7.8, LOS: 99.1 %, DrawRatio: 22.2 %
    SPRT: llr 2.2 (100.1%), lbound -2.2, ubound 2.2 - H1 was accepted

3) -20 doubled pawn penalty:
      Score of luna_dev vs luna_main: 2594 - 2428 - 1383  [0.513] 6405
    ...      luna_dev playing White: 1304 - 1209 - 690  [0.515] 3203
    ...      luna_dev playing Black: 1290 - 1219 - 693  [0.511] 3202
    ...      White vs Black: 2523 - 2499 - 1383  [0.502] 6405
    Elo difference: 9.0 +/- 7.5, LOS: 99.0 %, DrawRatio: 21.6 %
    SPRT: llr 2.2 (100.2%), lbound -2.2, ubound 2.2 - H1 was accepted
4) null move pruning:
      Score of luna_dev vs luna_main: 251 - 133 - 119  [0.617] 503
    ...      luna_dev playing White: 126 - 68 - 58  [0.615] 252
    ...      luna_dev playing Black: 125 - 65 - 61  [0.620] 251
    ...      White vs Black: 191 - 193 - 119  [0.498] 503
    Elo difference: 83.1 +/- 27.1, LOS: 100.0 %, DrawRatio: 23.7 %
    SPRT: llr 2.2 (100.1%), lbound -2.2, ubound 2.2 - H1 was accepted
5) reverse futility pruning:
      Score of luna_dev vs luna_main: 215 - 105 - 125  [0.624] 445
    ...      luna_dev playing White: 104 - 53 - 66  [0.614] 223
    ...      luna_dev playing Black: 111 - 52 - 59  [0.633] 222
    ...      White vs Black: 156 - 164 - 125  [0.491] 445
    Elo difference: 87.7 +/- 27.9, LOS: 100.0 %, DrawRatio: 28.1 %
    SPRT: llr 2.2 (100.1%), lbound -2.2, ubound 2.2 - H1 was accepted
6) razoring:
      Score of luna_dev vs luna_main: 303 - 195 - 239  [0.573] 737
    ...      luna_dev playing White: 149 - 105 - 115  [0.560] 369
    ...      luna_dev playing Black: 154 - 90 - 124  [0.587] 368
7) use TT lookup in qsearch: (some improvement; match statistics lost)

--------------------------------------------------
quick match against VICE (~2050 elo):

Score of luna_dev vs vice: 176 - 42 - 22  [0.779] 240
...      luna_dev playing White: 91 - 23 - 6  [0.783] 120
...      luna_dev playing Black: 85 - 19 - 16  [0.775] 120
...      White vs Black: 110 - 108 - 22  [0.504] 240
Elo difference: 219.0 +/- 49.9, LOS: 100.0 %, DrawRatio: 9.2 %
SPRT: llr 2.2 (100.1%), lbound -2.2, ubound 2.2 - H1 was accepted

quick match against zevra 1.8.6 (2269 elo):

Score of luna_dev vs zevra: 532 - 391 - 134  [0.567] 1057
...      luna_dev playing White: 263 - 194 - 72  [0.565] 529
...      luna_dev playing Black: 269 - 197 - 62  [0.568] 528
...      White vs Black: 460 - 463 - 134  [0.499] 1057
Elo difference: 46.6 +/- 19.7, LOS: 100.0 %, DrawRatio: 12.7 %
SPRT: llr 2.2 (100.3%), lbound -2.2, ubound 2.2 - H1 was accepted
--------------------------------------------------

8) futility pruning:
      Score of luna_dev vs luna_main: 50 - 37 - 39  [0.552] 126
    ...      luna_dev playing White: 20 - 21 - 23  [0.492] 64
    ...      luna_dev playing Black: 30 - 16 - 16  [0.613] 62
    ...      White vs Black: 36 - 51 - 39  [0.440] 126
    Elo difference: 36.0 +/- 50.9, LOS: 91.8 %, DrawRatio: 31.0 %
    SPRT: llr 0.253 (11.5%), lbound -2.2, ubound 2.2

--------------------------------------------------
quick match against teki 2 (2439 elo):

Score of luna_dev vs teki: 551 - 411 - 150  [0.563] 1112
...      luna_dev playing White: 269 - 224 - 64  [0.540] 557
...      luna_dev playing Black: 282 - 187 - 86  [0.586] 555
...      White vs Black: 456 - 506 - 150  [0.478] 1112
Elo difference: 44.0 +/- 19.1, LOS: 100.0 %, DrawRatio: 13.5 %
SPRT: llr 2.2 (100.2%), lbound -2.2, ubound 2.2 - H1 was accepted
--------------------------------------------------

9) added first late-move depth reduction (classic LMR):
      Score of luna_dev vs luna_main: 1625 - 1488 - 1471  [0.515] 4584
    ...      luna_dev playing White: 832 - 720 - 741  [0.524] 2293
    ...      luna_dev playing Black: 793 - 768 - 730  [0.505] 2291
    ...      White vs Black: 1600 - 1513 - 1471  [0.509] 4584
    Elo difference: 10.4 +/- 8.3, LOS: 99.3 %, DrawRatio: 32.1 %
    SPRT: llr 2.21 (100.4%), lbound -2.2, ubound 2.2 - H1 was accepted

10) added second late-move depth reduction (extra reduction in non-PV lines)
      Score of luna_dev vs luna_main: 367 - 279 - 345  [0.544] 991
    ...      luna_dev playing White: 192 - 131 - 173  [0.561] 496
    ...      luna_dev playing Black: 175 - 148 - 172  [0.527] 495
    ...      White vs Black: 340 - 306 - 345  [0.517] 991
    Elo difference: 30.9 +/- 17.5, LOS: 100.0 %, DrawRatio: 34.8 %
    SPRT: llr 1.79 (81.7%), lbound -2.2, ubound 2.2

11) added third late-move depth reduction (extra reduction if > 3 quiets
    searched and null move pruning failed): FAILED

12) added fourth late-move depth reduction (extra reduction if not improving):
      Score of luna_dev vs luna_main: 873 - 761 - 1068  [0.521] 2702
    ...      luna_dev playing White: 431 - 373 - 548  [0.521] 1352
    ...      luna_dev playing Black: 442 - 388 - 520  [0.520] 1350
    ...      White vs Black: 819 - 815 - 1068  [0.501] 2702
    Elo difference: 14.4 +/- 10.2, LOS: 99.7 %, DrawRatio: 39.5 %
    SPRT: llr 2.21 (100.4%), lbound -2.2, ubound 2.2 - H1 was accepted

13) added fifth late-move depth reduction (reduction TO the reduction if this
    move is a queen promotion): taking too long, will retest later

14) fixed material draw bug that would make illegal moves, as well as the other
    illegal move bug:
        Score of luna_dev vs luna_main: 1133 - 1014 - 1380  [0.517] 3527
  ...      luna_dev playing White: 491 - 506 - 767  [0.496] 1764
  ...      luna_dev playing Black: 642 - 508 - 613  [0.538] 1763
  ...      White vs Black: 999 - 1148 - 1380  [0.479] 3527
  Elo difference: 11.7 +/- 8.9, LOS: 99.5 %, DrawRatio: 39.1 %

defs.h
#ifndef DEFS_H
#define DEFS_H

#include <stdint.h>

typedef uint64_t U64;

// general definitions:
#define NAME "luna"
#define AUTHOR "ori yonay"
#define VERSION "0.0.1"
#define MAX_POSITION_MOVES 219
#define MAX_GAME_MOVES 512
#define DEFAULT_TT_SIZE 32 // in MB
#define DEFAULT_UCI_INPUT_BUFFER_SIZE 4096

#define MATE_IN_MAX 999970 // INF - 30
#define INF 1000000
#define MAX_SEARCH_PLY 64
#define NO_SCORE 10000000 // too high to be reached, so it's safe to use this

#define FEN_START "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"

// transposition table definitions:
#define TT_EXACT 0
#define TT_ALPHA 1
#define TT_BETA 2
#define TT_NO_MATCH 2147483600

// game phase enums:
enum {OPENING_PHASE, ENDGAME_PHASE};

// these are set up in a clever way so that we can add them up -
// for example, WHITE + KNIGHT = WN, BLACK + ROOK = BR, etc.
enum {WHITE, BLACK = 6};
enum {PAWN, KNIGHT, BISHOP, ROOK, QUEEN, KING};
enum {WP, WN, WB, WR, WQ, WK, BP, BN, BB, BR, BQ, BK, NONE};

// enum for board squares:
enum {
  A8, B8, C8, D8, E8, F8, G8, H8,
  A7, B7, C7, D7, E7, F7, G7, H7,
  A6, B6, C6, D6, E6, F6, G6, H6,
  A5, B5, C5, D5, E5, F5, G5, H5,
  A4, B4, C4, D4, E4, F4, G4, H4,
  A3, B3, C3, D3, E3, F3, G3, H3,
  A2, B2, C2, D2, E2, F2, G2, H2,
  A1, B1, C1, D1, E1, F1, G1, H1
};

// macro for files (ex. the A file bitmask can be found at FILES[A])
enum {A, B, C, D, E, F, G, H};

// macros for initial king and rook locations (used to update castle rights):
// LWR = left white rook, RBR = right black rook, etc.
enum {INIT_WK = 0x1000000000000000L, INIT_BK = 0x10L};

// utility macro for computing index of least significant bit:
#define LSB(x) (DEBRUIJN_INDEX[((x ^ (x-1)) * DEBRUIJN) >> 58])
// #define LSB(x) __builtin_ctzll(x) // <-- no noticeable performance difference

// utility macro for popping the least significant bit:
#define POP_LSB(x) (x &= x - 1)

// utility to isolate LSB:
#define ISOLATE_LSB(x) (x & ~(x - 1))

// are there multiple ones set in this bitboard:
#define SEVERAL(x) (x & (x - 1))

// number of set bits in the bitboard:
#define POPCOUNT(x) (__builtin_popcountll(x))

// macros for getting information out of a move integer:
#define MOVE_TO(move) ((move >> 26) & 0x3F)
#define MOVE_FROM(move) ((move >> 20) & 0x3F)
#define MOVE_CAPTURED(move) ((move >> 16) & 0xF)
#define MOVE_IS_EP(move) (move & 0x8000)
#define MOVE_IS_PAWNFIRST(move) (move & 0x4000)
#define MOVE_IS_CASTLE(move) (move & 0x2000)
#define MOVE_IS_PROMOTION(move) (move & 0x1000)
#define MOVE_PROMOTION_PIECE(move) ((move >> 8) & 0xF)
#define MOVE_PIECEMOVED(move) (move & 0xF)
#define MOVE_PCR(move) ((move >> 4) & 0xF)

// macros for determining castle rights:
// CWK = castle white kingside, CBQ = castle black queenside, etc.
#define CAN_CWK(castle_rights) (castle_rights & 0x8)
#define CAN_CWQ(castle_rights) (castle_rights & 0x4)
#define CAN_CBK(castle_rights) (castle_rights & 0x2)
#define CAN_CBQ(castle_rights) (castle_rights & 0x1)

// macros for getting the file and rank indices from a board index:
#define FILE_NO(idx) (idx % 8)
#define RANK_NO(idx) (8 - (idx / 8)) // +1 for human indexing

// which piece is it? (ex. PIECE_TYPE(WQ) = QUEEN)
#define PIECE_TYPE(piece) (piece % 6)

// defines for making castling code more readable:
#define CWK_ROOK_MASK 0xA000000000000000L
#define CWQ_ROOK_MASK 0x900000000000000L
#define CBK_ROOK_MASK 0xA0L
#define CBQ_ROOK_MASK 0x9L

#endif
